feat(TS): make nullable fields use option `?`

willemneal opened the pull request:
Now fields and type aliases that are nullable are encoded as `T?` instead of `T | null`.

For fields (properties), this means `T | undefined` and for type aliases `T | null` as before.

alexcrichton commented:
Thanks! I was gonna comment that I think it'd be good to have some tests for this as well, but it looks like the test updates to get CI green may suffice for that.

willemneal commented:
@alexcrichton I fixed the tests.  I apologize in advance for the code that unwraps a nullable type.  The update also fixed a bug in one of the tests when trying to access a method on a potentially undefined object.

alexcrichton commented on crates/gen-js/src/lib.rs:
```diff
@@ -2470,3 +2488,18 @@ impl Source {
     }
 }
+
+fn is_nullable(iface: &Interface, ty: &Type) -> bool {
```
Could this perhaps return `Option<&Type>` so the match here doesn't have to be redone above?

alexcrichton replied:
Also, would it be possible to remove this method and use `self.is_nullable_option` instead? (or some sort of shared helper method)

willemneal replied:
Before I blindly copied and didn't read too much into the details, but now I'm a little confused.

```rust
match variant.as_option() {
        Some(ty) => match ty {
              Type::Id(id) => match &iface.types[*id].kind {
                  TypeDefKind::Variant(v) => !self.is_nullable_option(iface, v),
                  ty => true,
              },
              _ => true,
          },
          None => false,
      }
```

Why is this not just `variant.as_option().is_some()`?  Wouldn't anything with `option<T>` be a nullable option, regardless of `T`?

willemneal replied:
Nvm I see.  `option<option<T>>` becomes `Option<T | null>` in TS.

alexcrichton commented on crates/gen-js/src/lib.rs:
```diff
@@ -189,7 +189,7 @@ impl Js {
                         if self.is_nullable_option(iface, v) {
                             self.print_ty(iface, v.cases[1].ty.as_ref().unwrap());
-                            self.src.ts(" | null");
+                            self.src.ts(" | null")
```
I think this is a stray change now?

willemneal replied:
No you still need it for cases like function signatures.  While you can have:
```ts
function foo(x?: string);
```

It has the requirement that the optional param comes after any non-optional ones, which isn't the case generally in wit, e.g. would fail

```ts
function foo(x?: string, y: number); 
function foo(x: string | null, y: number);  // <--- forces caller to put null
```

Perhaps if wit makes the same requirement about option types we could remove this.

alexcrichton replied:
I mean moreso the drop of the semicolon

alexcrichton commented on crates/gen-js/src/lib.rs:
```diff
@@ -140,6 +140,16 @@ impl Js {
         }
     }
 
+    fn unwrap_nullable_option<'a>(&self, iface: &'a Interface, ty: &Type) -> Option<&'a Type> {
```
Instead of duplicating the logic fo `is_nullable_option` could there be only one method here? (probably this one) and callers of `is_nullable_option` switch to this method with a `.is_some()` or something like that?

Also I think this may be better named something like `get_nullable_option` since "unwrap" typically has panicking connotations in Rust.

willemneal replied:
I moved the methods into the Interface struct, which opens the door for more refactoring in a later PR.

alexcrichton commented on crates/parser/src/lib.rs:
```diff
@@ -465,6 +465,27 @@ impl Interface {
+            self.get_nullable_option(ty)
+                .map_or(true, |ty| self.get_nullable_option(ty).is_none())
+        })
+    }
```
I don't think that this should live in the `wit-parser` crate because this is a JS-specific operation. Options are represented differently in each language and each one may need its own special casing.

willemneal replied:
While I agree that different languages will represent them differently, this allows any language to check if the type is `option<T>` vs `option<option<T>>`, which isn't about representation but rather a fact about the variant.

alexcrichton replied:
Perhaps, yeah, but I would personally prefer that this remain language-specific for now and in the future if there's duplication between languages we can deduplicate at that time.

syrusakbary commented:
This is great!

Would this PR make the following possible? (like in Typescript):

```
record person {
  name: string,
  age?: u32,
}
```

willemneal commented:
@syrusakbary  Yep!
```
record person {
  name: string,
  age: option<u32>
}
```
becomes
```ts
interface person {
  name: string,
  age?: number
}

alexcrichton commented:
Thanks!

syrusakbary commented:
Niice! Love it!
